### ISSUE TITLE
Memory allocation error: block too big

### ISSUE THREAD

**Issue by DimaSheiko** [ajseward/TradeSkillMaster#13]
This happens when using GetAllScan and there are more than 100k items in the auction which translates to more than 2k pages.

Is it possible to fix this error from the addon? Or is it a limitation at client or language level?
Because for example Addons like Auctioneer are able to scan the auction with a GetAllScan. I guess Auctioneer will make some kind of strategy to avoid memory overflow. And to do that in TSM you will have to rewrite a lot of code.

```
Addon: TradeSkillMaster
Message:  Interface\AddOns\TradeSkillMaster\Core\Threading.lua:28: memory allocation error: block too big
Date: 12/13/21 10:06:25
Client: 3.3.5
Locale: enEN
Stack:
[C]: <pcall>
[string "*:OnLoad"]:4: <[string "*:OnLoad"]:1>
[C]: <error>
TSM\Core\Threading.lua:28: <callback>
TSM\Util\Delay.lua:30: <Delay.lua:18>
TSM Stack:
?
Local Variables:

TSM Event Log:

Addons:
    Auctionator (2.6.3)
    AutoFlood ()
    BuyEmAll (2.9.2)
    CombatLogFix ()
    Details_3DModelsPaths (r267)
    Details ()
    Details_DeathGraphs ()
    Details_EncounterDetails ()
    Details_DataStorage ()
    Details_SunderCount ()
    Details_TimeLine ()
    Details_TinyThreat ()
    Details_ChartViewer ()
    Doom_CooldownPulse ()
    Elephant (2.9.6-beta)
    GearScoreLite (3x04)
    LoseControl (3.33)
    Mapster (1.3.9)
    NiceDamage (40000)
    Overachiever (v0.56)
    PallyPower (v3.2.20)
    PlateBuffs (1.5.4beta)
    Postal (v3.3.2)
    PowerAuras (3.0.0S)
    RaidRoll ()
    RaidRoll_LootTracker ()
    TidyPlates (5.13.3 (R39))
    WeakAuras (3.2.3)
    WeakAurasArchive ()
    WeakAurasModelPaths (2.16.0-beta1)
    WeakAurasOptions (3.2.3)
    WIM (3.3.7)
    ZoneLevelInfo (1.1)
    TSM (v2.8.3)
    TSM_Accounting (v2.3)
    TSM_AuctionDB (v2.3.10)
    TSM_Auctioning (v2.3.3)
    TSM_Crafting (v2.5.2)
    TSM_Destroying (v2.1.1)
    TSM_ItemTracker (v2.0.8)
    TSM_Mailing (v2.2.1)
    TSM_Shopping (v2.3.6)
    TSM_Warehousing (v2.0.10)
    ElvUI (6.08)
    ElvUI_ProjectZidras (2.0)
    ElvUI_AddOnSkins (1.06)
    ElvUI_BagControl (2.2)
    ElvUI_CustomTweaks (1.30)
    ElvUI_Enhanced (1.05)
    ElvUI_ImprovedSpecSwitch (1.00)
    ElvUI_MicrobarEnhancement (1.77)
    ElvUI_OptionsUI (1.06)
    ElvUI_RaidMarkers (4.24)
    ElvUI_VisualProcs (1.03)
    DBM-Core ()
```

**Comment by SorteKanin** [comment]
Got this too but with a perhaps more specific and useful error. I'm not sure how to fix this myself but any pointers would be appreciated.

```
Date: 2022-02-27 12:05:49
ID: -1
Error occured in: Global
Count: 2
Message: memory allocation error: block too big
Debug:
   (tail call): ?
   (tail call): ?
   [C]: tinsert()
   ...dOns\TradeSkillMaster_AuctionDB\Modules\Scanning.lua:198:
      ...dOns\TradeSkillMaster_AuctionDB\Modules\Scanning.lua:183
   ...erface\AddOns\TradeSkillMaster\Core\ErrorHandler.lua:406:
      ...erface\AddOns\TradeSkillMaster\Core\ErrorHandler.lua:404
   (tail call): ?
   ...dOns\TradeSkillMaster_AuctionDB\Modules\Scanning.lua:25: callbackHandler()
   ...\AddOns\TradeSkillMaster\Auction\AuctionScanning.lua:28:
      ...\AddOns\TradeSkillMaster\Auction\AuctionScanning.lua:26
   (tail call): ?
   ...erface\AddOns\TradeSkillMaster\Core\ErrorHandler.lua:406:
      ...erface\AddOns\TradeSkillMaster\Core\ErrorHandler.lua:404
   (tail call): ?
   ...\AddOns\TradeSkillMaster\Auction\AuctionScanning.lua:44:
      ...\AddOns\TradeSkillMaster\Auction\AuctionScanning.lua:32
   [string "safecall Dispatcher[1]"]:4:
      [string "safecall Dispatcher[1]"]:4
   [C]: ?
   [string "safecall Dispatcher[1]"]:13: ?()
   ...ter\Libs\CallbackHandler-1.0\CallbackHandler-1.0.lua:90: Fire()
   ...\TradeSkillMaster\Libs\AceEvent-3.0\AceEvent-3.0.lua:120:
      ...\TradeSkillMaster\Libs\AceEvent-3.0\AceEvent-3.0.lua:119
AddOns:
  Swatter, v3.1.14 (<%codename%>)
  Babylonian, v5.1.DEV.130
  Configator, v5.1.DEV.190
  Gatherer, v3.1.14
  GearScoreLite, v3x04
  LibExtraTip, v5.12.DEV.350(/embedded)
  SlideBar, v3.1.14 (<%codename%>)
  TradeSkillMaster, vv2.8.3
  TradeSkillMasterAccounting, vv2.3
  TradeSkillMasterAuctionDB, vv2.3.10
  TradeSkillMasterAuctioning, vv2.3.3
  TradeSkillMasterCrafting, vv2.5.2
  TradeSkillMasterDestroying, vv2.1.1
  TradeSkillMasterItemTracker, vv2.0.8
  TradeSkillMasterMailing, vv2.2.1
  TradeSkillMasterShopping, vv2.3.6
  TradeSkillMasterWarehousing, vv2.0.10
  BlizRuntimeLib_enUS v3.3.5.30300 <logon.warmane.com>
  (ck=271)

```

**Comment by ajseward** [comment]
So this is kinda an issue with the wow client itself. It's essentially found so many of the same item it's exceeding the maximum memory allowed for a table in lua.

limited it to only add up to 5000 of a single item, dunno if thats an appropriate number so feel free to re-open.

**Comment by Bananaman** [comment]
@andrew6180 It still happens with the latest code... on Warmane's Icecrown.

```
Addon: TradeSkillMaster
Message:  Interface\AddOns\TradeSkillMaster\Core\Threading.lua:28: memory allocation error: block too big
Date: 03/14/23 22:10:45
```

I wonder if there's another solution. I noticed that it builds a huge table of items and then sends it into `TSM.Data:ProcessData(data)`.

Would it be possible to somehow keep track of how many item IDs have been added to `data` so far, and let's say every 500 items, call `TSM.Data:ProcessData(data)` and reset data to an empty table, basically repeatedly chunking the scan results? Is that at all doable or would it break stuff?

**Comment by Bananaman** [comment]
2000 still gives error too. Nah man, this "fix" doesn't work. :D

**Comment by Bananaman** [comment]
Well I found why this "fix" has no effect. The `Scan.ProcessGetAllScan(self)` runs when doing a fast "Get All" scan (instead of waiting hours for a slow "normal scan".

There is no limiting or anything applied to the `Scan.ProcessGetAllScan(self)` function, so of course this fix has no effect.

**Comment by Bananaman** [comment]
I have found the reason it runs out of memory. Holy shit TSM is one of the shittiest addons I've ever seen. The original author must have ZERO IQ.

![image](https://user-images.githubusercontent.com/38923130/225164019-5379efef-b0a2-4253-9922-152e2ae6a6e2.png)

That's it. This line in the "process getall scan" does the following:

- Add 1 table entry for every individual item in a stack.
- Now imagine that the auction house contains a stack of 1000 "Arrows".
- This will add 1000 table rows, one for every arrow. And will mark the price of each row as the per-arrow buyout price.
- Now imagine that the auction house contains 500 stacks of 1000 arrows, that means it adds 500 000 (HALF A MILLION) entries to the "arrows" table.

This is UNFIXABLE. It would require a total rewrite of how TSM's code calculates prices when processing data, so that it could store 1 entry per auction/stack, instead of 1 entry per item per stack.

This is literally braindead code. It means extremely slow processing, extremely bloated memory, extreme slowdown, and of course we hit the memory cap when the code is this fucking stupid.

Sorry there isn't anything we can do about it. We would have to rewrite the code of TSM's "ProcessData" function to accept smarter, more compact tables of data. And this would be a major rewrite of all parts of TSM. It's not something I am willing to do.

I have looked at the competition btw:

- Aux: Not good. Doesn't support "GetAll" scan. But apart from that, it's able to do pretty advanced pricing algorithms, like TSM.
- Auctioneer and Auctionator: Both are pretty basic compared to TSM, sadly.

At this stage, maybe I lean towards rewriting Aux (https://gitlab.com/knights-of-sunwell/aux-addon/-/commits/master) to support "GetAll" scan, but I dunno, right now I wanna drop all this shit and just go enjoy the game.

**Comment by DimaSheiko** [comment]
@Bananaman Thanks for taking the time to analyze all that, I've been digging into TSM for a long time too, modifying things that I personally need, but I haven't touched anything regarding the price calculations.

Since I saw the limitation that andrew made I knew that was not the solution for the same reasons you said. Besides that, this way you will not limit the amount of items that come in auction but still the calculations will not be accurate, because if the auction has more than 55000 posts it will not give you all the items that are in auction unless you scan it manually, that is a problem because for example in warmane there are almost always more than 100000 posts in auction, so in theory you are asking for half of the items that are in auction if you don't do a manual full scan.

Since TSM V4 all the code was rewritten from 0 leaving behind the spagetti code you mention (or at least that's what I imagine). Maybe by taking a look at the new versions of TSM you could see how it works there now. Anyway it is a completely different code and very incompatible with the old 3.3.5a wow version so just check it out if you are curious.

**Comment by Bananaman** [comment]
@DimaSheiko Yeah I used TSM 4+ on retail and it was a great addon. I was shocked to see this old and braindead spaghetti code in the old TSM we have here.

I've done about half of the required analysis now, and I've changed the "GetAll" scanner to store the data in the more compact format, and have begun the work of rewriting the "market value" algorithm to support using the compact data tables (1 table entry per auction/stack, instead of millions of table entries for every individual item of every stack).

What remains to do:

- Rewrite the market value algorithm to calculate market value via the compact data. This might be very hard, the algorithm is super messy. But I think it can be done.
- Make the other scanners (not just GetAll) use this new compact data format.
- Look all over the codebase for `.records` usage to make sure we only generate compact everywhere else too.
- Some testing (which is hard because Warmane has a 30 minute cooldown between GetAll scans, sigh...).

But yeah it seems doable, and I might be able to have a pull request in a few days, if all goes well.

**Comment by DimaSheiko** [comment]
The algorithm is not so difficult to understand to do the market price calculation, I did the same TSM algorithm in PHP, I did it to generate the file that generates the original TSM Application, so that the addon could take that file and process it as it works in the official wow server. You can find that code [here](https://github.com/Tauri-WoW-Community-Devs/TauriTSMAppServer/blob/master/app/Http/Controllers/GenerateFileController.php). 

If you want a more detailed step by step explanation of what you should do you can find it [here](https://support.tradeskillmaster.com/en_US/custom-strings/how-is-auctiondb-market-value-calculated).

**Comment by Bananaman** [comment]
@DimaSheiko Yeah when I said the algorithm is messy, I meant that it's messy. This is the code:

![image](https://user-images.githubusercontent.com/38923130/225173747-39fecb2e-25c8-4729-83be-b6361c5822ee.png)

Super badly named variables, really stupid things like calculating values inside if-statements nad comparing to other calculated values in the if-statements, etc.

I will have to rewrite this to be able to handle the compact data. I have barely even started looking at this algorithm yet since I am tired after 2 hours of chasing through the bad code. :)

But I am sure it can be rewritten, it will just need some sleep and then I can look at it again.

**Comment by Bananaman** [comment]
@DimaSheiko Btw I had a look in the latest TSM4, and they don't seem to have the algorithm in the addon anymore. I suppose all "dbmarket" nowadays is calculated on their servers and downloaded to the addon via their software. Makes sense. So I will be forced to rewrite the algorithm above, unfortunately.

If you wanna mess around with it a bit, here's my current code. The else-statement is where we should be implementing the new algorithm:

```lua
function Data:CalculateMarketValue(data)
	local records = data.records  -- All separated auction stacks for this item (contains price per item and stack count).
	local quantity = data.quantity  -- How many of this item currently exist in total on the auction house (combines all stacks).

	if #records > 0 and type(records[1]) ~= "table" then
		-- USE THE OLD, SHITTY, BRAINDEAD ALGORITHM IF WE'VE BEEN GIVEN OLD-SCHOOL "BLOATED" RECORDS.
		TSM:Print("Warning: Calculating old-school market value. The code needs fixing.")  -- TODO
		local totalNum, totalBuyout = 0, 0
		local numRecords = #records
		
		for i=1, numRecords do
			totalNum = i - 1
			if i ~= 1 and i > numRecords*MIN_PERCENTILE and (i > numRecords*MAX_PERCENTILE or records[i] >= MAX_JUMP*records[i-1]) then
				break
			end
			
			totalBuyout = totalBuyout + records[i]
			if i == numRecords then
				totalNum = i
			end
		end
		
		local uncorrectedMean = totalBuyout / totalNum
		local variance = 0
		
		for i=1, totalNum do
			variance = variance + (records[i]-uncorrectedMean)^2
		end
		
		local stdDev = sqrt(variance/totalNum)
		local correctedTotalNum, correctedTotalBuyout = 1, uncorrectedMean
		
		for i=1, totalNum do
			if abs(uncorrectedMean - records[i]) < 1.5*stdDev then
				correctedTotalNum = correctedTotalNum + 1
				correctedTotalBuyout = correctedTotalBuyout + records[i]
			end
		end
		
		local correctedMean = floor(correctedTotalBuyout / correctedTotalNum + 0.5)
		
		return correctedMean
	else
		TSM:Print("Calculating intelligent market value for all items...")
		local qq = 0
		for idx=1, #records do
			local stack_size, buyout_per_item = unpack(records[idx])
			qq = qq + stack_size
		end

		TSM:Print(format("Quantity: %d, Calculated Quantity: %d", quantity, qq))

		return 0  -- TODO: Screws up the market value for now but avoids errors...
	end
end
```

The two variables we have to work with are listed at the top of the function:

- `records` = Table of 1 auction per row (this is the new compact format). The rows are formatted as follows: `tinsert(data[itemID].records, {stack_size, buyout_per_item})` where `stack_size` is how many items are in that stack (such as 980) and `buyout_per_item` is the price of every individual item in that stack.
- `quantity` = The total count of items for that item ID, let's say there's 5 stacks of 1000 arrows each, in that case `quantity` will be `5000`, and `records` will contain 5 rows that each consist of `{1000, <some price per item here>}`.

**Comment by Bananaman** [comment]
Aight I was able to capture a "GetAll" result into a SavedVariable which will speed up development tomorrow. I'll be able to work on the algorithm without having to wait 30 minutes between "GetAll" throttled scans. :)

For fun, here's some statistics. I made it print the total of all items on the auction house. This is how many "table rows" the previous code generated:

![image](https://user-images.githubusercontent.com/38923130/225180604-7dcc0dd2-a2f2-4d54-9caf-c231db536d6d.png)

And here's how many table rows my new code generates:

![image](https://user-images.githubusercontent.com/38923130/225180805-2beeb748-ad50-4368-838c-666499c6d074.png)

In other words, the new algorithm will only use 4.6% as much memory as the old one. :) Well, enough. Time to sleep.

**Comment by Bananaman** [comment]
@DimaSheiko Hmm, forget it. Their algorithm does a bunch of forwards and backwards lookups in the array. I've decided to do a compromise solution which still solves the memory issue!

- I store everything as compact "records" when doing the actual scan.
- The market price calculation algorithm processes 1 item (all of its records) at a time. So I will simply make that algorithm "expand" the compact per-item records into one big, bloated table for ONLY THAT ITEM, and then process it using their old, messy algorithm exactly as-is, and then we simply delete the temporary table.
- I have already added some debug testing to ensure that this will work. I made it print every item that would generate over 5k "table record rows". It isn't many items at all. Most items would generate small tables. The issue in their old code was that *EVERY* item in the auction house generated bloated tables all simultaneously. My new algorithm will only do that per-item when each item is being processed. I think this pull request will be ready tomorrow.

Here's the statistics for items that would generate over 5000 rows each, on warmane's icecrown, which you can see isn't many items:

```

TSM_AuctionDB: Triggering fake scan...
TSM_AuctionDB: BIG Quantity (>5k): 12000
TSM_AuctionDB: BIG Quantity (>5k): 22000
TSM_AuctionDB: BIG Quantity (>5k): 8867
TSM_AuctionDB: BIG Quantity (>5k): 169000
TSM_AuctionDB: BIG Quantity (>5k): 686900
TSM_AuctionDB: BIG Quantity (>5k): 9800
TSM_AuctionDB: BIG Quantity (>5k): 5817
TSM_AuctionDB: BIG Quantity (>5k): 10694
TSM_AuctionDB: BIG Quantity (>5k): 8606
TSM_AuctionDB: BIG Quantity (>5k): 26048
TSM_AuctionDB: BIG Quantity (>5k): 5137
```

**Comment by Bananaman** [comment]
@DimaSheiko Alright don't worry about it. I am almost done rewriting the algorithm to support chunking. You can celebrate tomorrow I think. :)

**Comment by Bananaman** [comment]
@DimaSheiko Success. Algorithm is rewritten and is ~3-5x faster than the old one, and uses around 800x LESS MEMORY now. You can start celebrating and anticipating the release. ;)

But I have to investigate some other things before I release this. I noticed that Warmane' "GetAll" returns only 55000 items even thoguh there are 122000 auctions on the AH. I'll see if there's any way to request "more/remaining" auctions, or if we have to settle for only being able to get 55k auctions via "GetAll" on huge private servers like that.

**Comment by DimaSheiko** [comment]
Congratulations, sorry I can't help you, I don't have much free time lately.
I tell you, it is only possible to proceed 55k auctions in the GetAll Scan due to the limitations, that's how blizzard created it, and it will be a problem to get the real market value.

**Comment by Bananaman** [comment]
@DimaSheiko 

> I tell you, it is only possible to proceed 55k auctions in the GetAll Scan due to the limitations, that's how blizzard created it, and it will be a problem to get the real market value.

Yeah that's what I suspected too. Thanks for the info! And it means that the market value calculations will not be correct for "GetAll" for big servers like that.

I have a solution... I will make it output a message in the GetAll scanner: If `total_auctions ~= getall_auctions` then say something like "TSM_AuctionDB: Your server has 122523 auctions, but we the server only gave us 55000. Please run normal "Full Scans" instead to fill in the missing market data properly."

**Comment by Bananaman** [comment]
@DimaSheiko Hahahahhhahah. I would have released the pull request yesterday, but I found a last-minute bug, which turns out to be... a bug... that is so big... that I can't believe ANYONE even used this addon. :D

The way this addon calculates market value when doing a Full Scan or Group Scan is 100% wrong. It generates prices like 500 gold for items worth 20 gold. :) I am pretty sure what the bug is caused by, but it means I will spend a little more time fixing that bug too in my spare time.

Just wanted to give the funny status update. This blew my mind.

**Comment by DimaSheiko** [comment]
lol I am not surprised, Maybe it is a problem similar to the one I discovered in the calculation of the minimum auction in the crafting queue. At the same time I would be afraid of you doing such a big rework because you could also make mistakes 😅.

**Comment by Bananaman** [comment]
> Maybe it is a problem similar to the one I discovered in the calculation of the minimum auction in the crafting queue.

It's similar to that kind of "stupid bug" yeah. This time, the bug is most likely caused by the auction scanner receiving the items in random order (like: per-item buyout values: 500g, 20g, 5g, 35g, 35g, 35g, etc), but the code of the market value algorithm REQUIRES that the data is sorted from lowest to highest price, so when the data is messy/unsorted like this, it totally screws up its analysis of rising prices, because it starts at like "the cheapest item costs 500g, hey wtf is this 20g item? that;'s a 480g price difference, so let's ignore the 20g item lol, yeah... this item is definitely worth 500g!". That's how the algorithm works (and since I rewrote it, I am sure about that). Therefore, my theory is that the auction scan data is coming in unsorted order, which would be the reason why it's screwing up the market calculations.

> At the same time I would be afraid of you doing such a big rework because you could also make mistakes .

Yeah I totally understand that fear, which is why it's funny that you said that. :D I have already implemented code which runs the market value calculations via the old shitty algorithm and my new (up to 7x faster and using only 5% as much RAM) algorithm, and ensures that the results are exactly the same. :)

![image](https://user-images.githubusercontent.com/38923130/225756066-d22ef7cf-1c70-4e34-bb46-254e705cd42e.png)

It helps that I've been a programmer for over 20 years. :) All is safe here. Gonna be giving you all a really nice update soon

**Comment by Bananaman** [comment]
Okay I didn't wanna have this "on my mind" constantly, so I went in and did a quick debug test by making the market value algorithm print every buyout price it looks at, in the order they appear in the input-data. Yep, TSM really was THAT badly broken before!

```
TSM_AuctionDB: --start, quantity: 270
TSM_AuctionDB: +new, price: 737500.0 <-- 73.75g netherweave bag
TSM_AuctionDB: +new, price: 85615363.0 <-- 8561.53 gold netherweave bag (!)
TSM_AuctionDB: +new, price: 259000.0 <-- normal ones for like 26g
TSM_AuctionDB: +new, price: 280000.0
TSM_AuctionDB: +new, price: 305500.0
TSM_AuctionDB: +new, price: 306060.0 <-- 30.6g
TSM_AuctionDB: +new, price: 737500.0
TSM_AuctionDB: +new, price: 210000.0
TSM_AuctionDB: +new, price: 216700.0 <-- 21.7g
TSM_AuctionDB: +new, price: 220000.0
TSM_AuctionDB: +new, price: 229900.0
TSM_AuctionDB: +new, price: 250000.0
TSM_AuctionDB: +new, price: 259000.0
TSM_AuctionDB: +new, price: 204000.0
TSM_AuctionDB: +new, price: 205000.0
TSM_AuctionDB: +new, price: 208800.0
TSM_AuctionDB: +new, price: 208900.0
TSM_AuctionDB: +new, price: 210000.0
TSM_AuctionDB: +new, price: 194900.0
TSM_AuctionDB: +new, price: 195000.0
TSM_AuctionDB: +new, price: 199000.0
TSM_AuctionDB: +new, price: 199500.0
TSM_AuctionDB: +new, price: 203000.0
TSM_AuctionDB: +new, price: 204000.0
TSM_AuctionDB: +new, price: 194900.0 <-- 19.5g
```

As you see, the input data to the algorithm is totally freaking unsorted, lmao. So the TSM algorithm sees the 73 gold netherweave bag, thinks "okay cool, the cheapest is 73 gold". Then it sees the 8000+ gold one and goes "nahh fuck this, that's too expensive" and stops analyzing prices.

Result: Even though most netherweave bags cost 19.5g, it thinks they are worth 73 gold as "market value".

:)

Glad to have verified the issue. There won't be any more status updates until I've fixed this and started the pull request. I think I'll do that tomorrow, since I was at the dentist today and need to relax. :) Take care for now!

**Comment by DimaSheiko** [comment]
Good catch! 
That's why [here](https://github.com/Tauri-WoW-Community-Devs/TauriTSMAppServer/blob/master/app/Http/Controllers/GenerateFileController.php#L55) I order it this way to avoid that problem

**Comment by Bananaman** [comment]
@DimaSheiko Yeah exactly, I will be changing the market price algorithm to always sort the data ("never trust the raw input").

That's a really great project, and great job reverse engineering how the TSM app's market Lua works. I love your clean code quality! :) I wish warmane had a public auction API too, so we could do the same on that server.
